### PR TITLE
HDDS-7466. SnapshotDiff Client Code.

### PR DESCRIPTION
**Pull request by nandakumar131**
## What changes were proposed in this pull request?

This jira adds SnapshotDiff client code and the server side handling of snapshotDiff RPC call.

## What is the link to the Apache JIRA

HDDS-7466

## How was this patch tested?

Manual testing and unit test.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java [hunk 1]
```diff
@@ -217,4 +224,40 @@ public static boolean isSnapshotKey(String[] keyParts) {
...
+
+  public SnapshotDiffReport getSnapshotDiffReport(final String volume,
+                                                  final String bucket,
+                                                  final String fromSnapshot,
+                                                  final String toSnapshot)
+      throws IOException {
+    // Validate fromSnapshot and toSnapshot
+    final SnapshotInfo fsInfo = getSnapshotInfo(volume, bucket, fromSnapshot);
+    final SnapshotInfo tsInfo = getSnapshotInfo(volume, bucket, toSnapshot);
+    verifySnapshotInfoForSnapDiff(fsInfo, tsInfo);
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:236]: We need to assert that the from and to snapshots are from same volume/bucket.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+    // TODO: Once RocksDBCheckpointDiffer exposes method to get list
+    //  of delta SST files, plug it in here.
+
+    Set<String> fromSnapshotFiles = RdbUtil.getKeyTableSSTFiles(fromSnapshot
+        .getMetadataManager().getStore().getDbLocation().getPath());
+    Set<String> toSnapshotFiles = RdbUtil.getKeyTableSSTFiles(toSnapshot
+        .getMetadataManager().getStore().getDbLocation().getPath());
+
+    final Set<String> deltaFiles = new HashSet<>();
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:61]: The sst files to be checked here need to be obtained from RocksDBCheckpointDiffer#getSSTDiffList . although we could keep this code path as fallback for cases where there is no DAG /compaction log optimisation

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/RdbUtil.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  private RdbUtil() { }
+
+  public static Set<String> getKeyTableSSTFiles(final String dbLocation)
+      throws RocksDBException {
+    final List<ColumnFamilyHandle> columnFamilyHandles  = new ArrayList<>();
+    final List<ColumnFamilyDescriptor> cfd = new ArrayList<>();
+    cfd.add(
+        new ColumnFamilyDescriptor(
+            "keyTable".getBytes(StandardCharsets.UTF_8)));
```
Comment (sadanand48) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/RdbUtil.java:49]: also need to consider file and directory tables.

**Comment by prashantpogde** [comment]
@nandakumar131  Can we resolve the conflicts ? I am planning to merge the changes back to master. It would be good if we can get these changes.

**Comment by nandakumar131** [comment]
This PR is not complete, it needs followup changes.

**Comment by prashantpogde** [comment]
LGTM. Lets file follow up Jiras for unit test + integration with the DAG code and we can merge this PR.

**Review by prashantpogde — APPROVED** [review]

**Pull request merged**